[Refactoring] A couple of fixes for async return handling

hamishknight opened the pull request:
- Fix placeholder handling so we don't crash while re-writing a return statement with a return expression
- Avoid duplicating return statements when converting a completion handler call into a return statement
- Don't drop return statements at the end of a block if they have an expression

rdar://77789360

hamishknight commented:
@swift-ci please test

bnbarham commented on lib/IDE/Refactoring.cpp:
```diff
@@ -5734,15 +5744,23 @@ class AsyncConverter : private SourceEntityWalker {
+      // It's possible the user has already written an explicit return statement
+      // for the completion handler call, e.g 'return completion(args...)'. In
+      // that case, be sure not to add another return.
+      auto *parent = getWalker().Parent.getAsStmt();
```
Getting sillier and sillier, but... what if we had `return (completion(arg))`? In that case the parent will be a `ParenExpr`. Not sure there's a better way to handle that than a flag on entering/leaving a `ReturnStmt` though.

I'm probably okay getting this in regardless since it fixes the most common case + the crashing.

hamishknight replied:
I *believe* that we won't transform `(completion(arg))` because we check for `NestedExprCount == 0`. So we'd likely need some extra handling to dig into the semantic expr to both transform it and then avoid adding an extra return. I can add a test either way to make sure we at least capture the current behaviour though

bnbarham replied:
Ah true - yeah, add test with FIXME I think.

hamishknight commented on include/swift/IDE/SourceEntityWalker.h:
```diff
@@ -176,6 +177,24 @@ class SourceEntityWalker {
   virtual void anchor();
+
+  /// Retrieve the current ASTWalker being used to traverse the AST.
+  const ASTWalker &getWalker() const {
```
I don't really mind either way, the walker is `const` so there's very little else you could do with it, but if it ever exposed additional state, that might be useful to be able to access.

hamishknight commented on lib/IDE/Refactoring.cpp:
```diff
@@ -5734,15 +5744,23 @@ class AsyncConverter : private SourceEntityWalker {
   void addHandlerCall(const CallExpr *CE) {
     auto Exprs = TopHandler.extractResultArgs(CE);
 
+    bool addedReturnOrThrow = true;
```
Ack, good catch!

hamishknight committed:
[Refactoring] Tweak handling of return and break placeholders

- Add a missing return to the break statement
placeholder handling.

- Only turn the `return` token into a placeholder,
as we still want to apply the transformation to
the sub expression.

This stops us from crashing by attempting to walk
into the return sub-expression.

rdar://77789360

hamishknight committed:
[Refactoring] Don't drop returns with expressions

Previously we were unconditionally dropping a
return statement if it was the last node, which
could cause us to inadvertently drop the result
expression as well. Instead, only drop bare
'return' statements.

rdar://77789360

hamishknight committed:
[IDE] Allow SourceEntityWalker to keep track of its ASTWalker

This will allow it to be queried for things like
the parent ASTNode in the current traversal.

hamishknight committed:
[test] Add test for async completion handler call in parens

We don't currently handle this, but arguably we
should be able to.

hamishknight committed:
[Refactoring] Avoid duplicating return statements

It's possible the user has already written an
explicit return for the call to the completion
handler. In that case, avoid adding another return.

rdar://77789360

hamishknight commented:
@swift-ci please smoke test

bnbarham commented:
Thanks for the extra test :)
